### PR TITLE
Bugfix FXIOS-10798 [Sponsored tiles] URLCache doesn't retrieve from cache

### PR DESCRIPTION
**Pull request by lmarceau**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10798)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/23562)

## :bulb: Description
URLCache doesn't retrieve from cache if there's an httpBody set on the request. I really hope we can set this code in fire soon, cause I don't like this fix at all 😆 

## :pencil: Checklist
You have to check all boxes before merging
- [X] Filled in the above information (tickets numbers and description of your work)
- [X] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/Frontend/Home/TopSites/DataManagement/UnifiedAds/UnifiedAdsProvider.swift [hunk 1]
```diff
@@ -114,7 +117,10 @@ class UnifiedAdsProvider: URLCaching, UnifiedAdsProviderInterface, FeatureFlagga
             guard let self = self else { return }
             switch result {
             case .success(let result):
-                self.cache(response: result.response, for: request, with: result.data)
+                // FXIOS-10798 - URLCache doesn't retrieve from cache if there's an httpBody set on the request
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Home/TopSites/DataManagement/UnifiedAds/UnifiedAdsProvider.swift:120]: Can we make a new ticket number for these notes to track setting this code on 🔥  in the new year, or whenever you thought we wouldn't need this code? 😆 (I don't have context on the project so maybe this whole file will get wiped, not sure lol)

**Review by ih-codes — COMMENTED** [review]
I'll defer to @issammani's approval since you two were looking at this today. It's wild! Just a question, do you ever need the cached request's body to do work for the Unified Ads? What is the benefit of caching the requests if there is no body? 🤔  Just curious how this code works since I don't have any context. 🙏

File: firefox-ios/Client/Frontend/Home/TopSites/DataManagement/UnifiedAds/UnifiedAdsProvider.swift (hunk 1 above)
Comment (lmarceau) [reply firefox-ios/Client/Frontend/Home/TopSites/DataManagement/UnifiedAds/UnifiedAdsProvider.swift:120]: At some point it is supposed to be wiped entirely, but... we never know lol. So if this feature gets integrated as is in the new homepage code, I'll note this down over there that this is not so great 😅 

The issue I have to track this work is [here](https://mozilla-hub.atlassian.net/browse/FXIOS-10757)

**Comment by lmarceau** [comment]
> Just a question, do you ever need the cached request's body to do work for the Unified Ads? 

No we don't need the request body apart from when we make the actual networking request. In the end it doesn't matter if it's not used to cache or to retrieve from the cache, but it does feel wrong as an identical request to cache and retrieve should give you the same result. 

> What is the benefit of caching the requests if there is no body?

It's the response to that request that we use to decode the sponsored tiles. So the data we want to retrieve is in fact the response. You can see it as the `request` being like a key to save and retrieve data from the cache.

**Review by issammani — APPROVED** [review]
This is a weird one. So I tried locally again in playground with a reduced test case and I can reproduce this behaviour. The fact that it caches it ( as evident by looking at cache size ) but it's not retrievable ( unless no body is present , we tried "foo".data() as well ) is weird. Looking further into `URLCache` [code](https://github.com/swiftlang/swift-corelibs-foundation/blob/b81f510d86933646026a731a92d043aea9238386/Sources/FoundationNetworking/URLCache.swift#L354-L371) the cache key doesn't seem to care about the body or make a distinction between a `GET` or `POST` ( this code is probably outdated though ). 

TL;DR I think this is a good fix seeing as we are doing everything we should but just can't seem to retrieve from cache when the body is cached as well. @lmarceau do you know if apple has something similar to webkit's bugzilla so maybe we can open a bug there ?

**Comment by lmarceau** [comment]
> @lmarceau do you know if apple has something similar to webkit's bugzilla so maybe we can open a bug there ?

I think we can report a bug through the [Feedback Assistant](https://developer.apple.com/bug-reporting/). There's a built in application in MacOS and we can create reports from there.

Screenshot 2024-12-12 at 9 19 50 AM
